src/libstore/rewrite-derivation.cc

```diff
@@ -0,0 +1,100 @@
+#include "rewrite-derivation.hh"
+#include "parsed-derivations.hh"
+#include "util.hh"
+
+namespace nix {
+
+void recomputeOutputs(Store & store, Derivation & drv) {
+    for (auto & i : drv.outputs) {
+        debug("Rewriting env var %s", i.first);
+        auto outputEnvVar = drv.env.find(i.first);
+        if (outputEnvVar != drv.env.end()) {
+            outputEnvVar->second = "";
+            debug("Rewrote env var %s", outputEnvVar->first);
+        }
+        i.second = DerivationOutput {
+            .path = StorePath::dummy,
+            .hash = std::optional<FixedOutputHash> {},
+        };
+    }
+
+    /* Use the masked derivation expression to compute the output path. */
+    Hash h = hashDerivationModulo(store, drv, true);
+
+    for (auto & i : drv.outputs) {
+        // XXX: There's certainly a better and less error-prone way
+        // of getting the name than to look it up in the drv environment
+        string name = ParsedDerivation(StorePath::dummy, drv).getStringAttr("name").value_or("");
+        StorePath outPath = store.makeOutputPath(i.first, h, name);
+        auto outputEnvVar = drv.env.find(i.first);
+        if (outputEnvVar != drv.env.end())
+            outputEnvVar->second = store.printStorePath(outPath);
+        debug(format("Rewrote output %1% to %2%")
```

Comment:
`format` is obsolete, please do `debug("rewrite output '%s' to '%s'", arg1, arg2)`.